### PR TITLE
[LValue Generalization] Add BoundsSiblingFields to CheckedCAnalysesPrepass [7/n]

### PR DESCRIPTION
**Pull request by kkjeer**
This PR modifies the CheckedCAnalysesPrepass to populate a map of fields `F` in a struct declaration `S` to the fields in `S` in whose declared bounds `F` appears.

We can use this map to synthesize AbstractSets for MemberExprs whose target bounds depend on a MemberExpr `M` that is being modified in an assignment. When synthesizing for a MemberExpr `s->f` (or `s.f`), we can check each sibling field `g` of `f` in whose declared bounds `f` appears, create a MemberExpr `s->g` (or `s.g`), and check whether `M` appears in the declared bounds of `g`.

### CODE REVIEW

File: clang/test/CheckedC/inferred-bounds/bounds-sibling-fields.c [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+  _Array_ptr<int> q : count(c->len);
+};
+
+struct A {
+  struct B *b;
+  int i;
+  int j;
+  _Array_ptr<int> start;
+  _Array_ptr<int> end;
+  _Array_ptr<int> p : count(b->c->len + b->i);
```
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/test/CheckedC/inferred-bounds/bounds-sibling-fields.c:26]: Yes, this is correct. I've expanded the comment in CheckedCAnalysesPrepass.h on `BoundsSiblingFields`.

The reason for this is that if a field `f` occurs in the declared bounds of an field `p` via a field access, this does not mean that we need to synthesize a member expression `s->p` when we call `SynthesizeMembers` on `s->f`. I'm working on a formalization for this claim so I can update the design document.

File: clang/lib/Sema/CheckedCAnalysesPrepass.cpp [hunk 2]
```diff
@@ -126,44 +221,62 @@ class PrepassHelper : public RecursiveASTVisitor<PrepassHelper> {
...
+      PrintDeclMap<const FieldDecl *>(FD, "BoundsSiblingFields",
+                                      Info.BoundsSiblingFields);
+    }
+
+    // Print a map from a key of type T to a set of elements of type T,
+    // where T should inherit from NamedDecl.
+    // This method can be used to print the BoundsVars and
+    // BoundsSiblingFields maps.
+    template<class T>
+    void PrintDeclMap(FunctionDecl *FD, const char *Message,
```
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/lib/Sema/CheckedCAnalysesPrepass.cpp:243]: With respect to synthesizing member expressions, I don't foresee a need to modify `BoundsSiblingFields` to distinguish between a field occurring in the lower vs upper bounds of another field. I agree that we could use this information when we extend bounds widening to support null-terminated array struct fields.

File: clang/lib/Sema/CheckedCAnalysesPrepass.cpp [hunk 3]
```diff
@@ -39,7 +40,83 @@ class PrepassHelper : public RecursiveASTVisitor<PrepassHelper> {
...
+    FieldDecl *FieldWithBounds = nullptr;
+
+    // ProcessedStructs keeps tracks of the struct declarations whose fields
+    // have been traversed by the FillBoundsSiblingFields method. This avoids
+    // unnecessary duplicate traversals of struct fields.
+    llvm::SmallPtrSet<const RecordDecl *, 2> ProcessedStructs;
+
+    // GetRecordDecl returns the struct declaration, if any, that is
+    // associated with the given type. For example, if the given type is
+    // struct S, struct S *, _Ptr<struct S>, etc., GetRecordDecl will
```
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/lib/Sema/CheckedCAnalysesPrepass.cpp]: Without the change from `if` to `while` in `GetRecordDecl`, the last test case in bounds-sibling-fields.c will fail. For example, if `T` is `_Array_ptr<_Ptr<struct S>>`, then `T->getPointeeOrArrayElementType()` will return `_Ptr<struct S>`. We need to keep calling `getPointeeOrArrayElementType` in order to get to the type `struct S`.

File: clang/lib/Sema/CheckedCAnalysesPrepass.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/lib/Sema/CheckedCAnalysesPrepass.cpp:258]: Based on #1050, it looks like separating `BoundsVars` into vars in lower and upper bounds is done via two maps: `BoundsVarsLower` and `BoundsVarsUpper`, both of which map `const VarDecl *` to a set of `const VarDecl *`. So it seems that this implementation of `PrintDeclMap` should work for `BoundsVarsLower` and `BoundsVarsUpper`. If we need to separate `BoundsSiblingFields` into lower and upper bounds in the future, we can use a similar approach.

File: clang/test/CheckedC/inferred-bounds/bounds-sibling-fields.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/test/CheckedC/inferred-bounds/bounds-sibling-fields.c:26]: I've updated the (internal) design document to include a discussion of this point. The essence of the argument is that, if the (concrete, expanded) bounds of a member expression `EBase->F` use the value of the modified member expression `MBase->MField`, where `EBase->EField` is the current argument `E` to `SynthesizeMembers`, then the (non-concrete, non-expanded) declared bounds of `F` must include a `DeclRefExpr` whose declaration is `EField`. When we concretize and expand the declared bounds of `F`, each `DeclRefExpr` `V` in the declared bounds of `F` is replaced with `EBase->V`.

File: clang/lib/Sema/CheckedCAnalysesPrepass.cpp [hunk 4]
```diff
@@ -39,7 +40,83 @@ class PrepassHelper : public RecursiveASTVisitor<PrepassHelper> {
...
+      // Do not traverse a struct declaration more than once.
+      // FillBoundsSiblingFields can be called more than once on a given
+      // struct declaration if a function declares multiple variables with
+      // the same struct type, or in case of recursive struct definitions.
+      if (ProcessedStructs.count(S))
+        return;
+
+      FieldWithBounds = nullptr;
+      ProcessedStructs.insert(S);
+      for (auto It = S->field_begin(), E = S->field_end(); It != E; ++It) {
```
[the remark this answers is not in the document]
Comment (kkjeer) [reply clang/lib/Sema/CheckedCAnalysesPrepass.cpp:103]: Created issue #1051 to track where clause support for bounds sibling fields.

**Review by mgrang — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**